google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerImpl.java

```diff
@@ -263,6 +265,19 @@ public void cancelled(Context context) {
         } else {
           backoffSleep(context, backOff);
         }
+        if (options.getRetrySettings() != null) {
```

Comment:
i would suspect the gax library being able to take care of these.

We should sync with gax|gapic team to help here and make this generic across all APIs and client libraries.

@hzyi-google - any thoughts on how to implement this?

Reply:
It does not seem this implementation is using any retry helpers provided by gax. what it's doing is extracting retry configurations set in `SpannerOptions` and does manual retries here.

Reply:
The Spanner library has its own implementation of runWithRetries(...), which also uses io.grpc.Context to allow the user to cancel calls or set timeouts. Re-writing this method, and thereby all the Spanner library, to use the standard helper methods from gax is probably a good idea. But as far as I can see, it will probably also have some side-effects to the current behavior.

Is that something we would want as part of solving this issue, or should that be a separate project?

Reply:
I agree with you that it will be ideal to use retry helpers in gax, and I also think it should be a separate project as well.

Reply:
@hzyi-google, should we close this PR and work on using the GAX retry helpers in another PR?

Reply:
@sduskis - we can have knut (or whoever) write up a design on migrating to gax, and its side-effects (or gaps in gax library) that Cloud Spanner may have a change in behavior.

I would start a new project and close this PR.
Also, there is a parallel thread for setting timeouts (and rpc cancellations) that should be included in this project (if possible) to make sure we dont drop any existing feature from the clients.

@crwilcox for timeouts & cancellations issue.